src/crypto/keccak.c

```diff
@@ -31,53 +31,116 @@ const uint64_t keccakf_rndc[24] =
+            st[i + 10] ^= t; \
+            st[i + 15] ^= t; \
+            st[i + 20] ^= t; \
         }
 
+        THETA(0);
+        THETA(1);
+        THETA(2);
+        THETA(3);
+        THETA(4);
+
         // Rho Pi
-        t = st[1];
-        for (i = 0; i < 24; i++) {
-            j = keccakf_piln[i];
-            bc[0] = st[j];
-            st[j] = ROTL64(t, keccakf_rotc[i]);
-            t = bc[0];
-        }
+        const uint64_t t = st[1];
+        st[ 1] = ROTL64(st[ 6], 44);
+        st[ 6] = ROTL64(st[ 9], 20);
+        st[ 9] = ROTL64(st[22], 61);
+        st[22] = ROTL64(st[14], 39);
+        st[14] = ROTL64(st[20], 18);
+        st[20] = ROTL64(st[ 2], 62);
+        st[ 2] = ROTL64(st[12], 43);
+        st[12] = ROTL64(st[13], 25);
+        st[13] = ROTL64(st[19],  8);
+        st[19] = ROTL64(st[23], 56);
+        st[23] = ROTL64(st[15], 41);
+        st[15] = ROTL64(st[ 4], 27);
+        st[ 4] = ROTL64(st[24], 14);
+        st[24] = ROTL64(st[21],  2);
+        st[21] = ROTL64(st[ 8], 55);
+        st[ 8] = ROTL64(st[16], 45);
+        st[16] = ROTL64(st[ 5], 36);
+        st[ 5] = ROTL64(st[ 3], 28);
+        st[ 3] = ROTL64(st[18], 21);
+        st[18] = ROTL64(st[17], 15);
+        st[17] = ROTL64(st[11], 10);
+        st[11] = ROTL64(st[ 7],  6);
+        st[ 7] = ROTL64(st[10],  3);
+        st[10] = ROTL64(t, 1);
 
         //  Chi
-        for (j = 0; j < 25; j += 5) {
-            for (i = 0; i < 5; i++)
-                bc[i] = st[j + i];
-            for (i = 0; i < 5; i++)
-                st[j + i] ^= (~bc[(i + 1) % 5]) & bc[(i + 2) % 5];
-        }
+        int j = 0;
+        bc[0] = st[j    ];
+        bc[1] = st[j + 1];
+
+        st[j    ] ^= (~st[j + 1]) & st[j + 2];
+        st[j + 1] ^= (~st[j + 2]) & st[j + 3];
+        st[j + 2] ^= (~st[j + 3]) & st[j + 4];
+        st[j + 3] ^= (~st[j + 4]) & bc[0];
+        st[j + 4] ^= (~bc[0]) & bc[1];
+
+        j = 5;
+        bc[0] = st[j    ];
+        bc[1] = st[j + 1];
+
+        st[j    ] ^= (~st[j + 1]) & st[j + 2];
+        st[j + 1] ^= (~st[j + 2]) & st[j + 3];
+        st[j + 2] ^= (~st[j + 3]) & st[j + 4];
+        st[j + 3] ^= (~st[j + 4]) & bc[0];
+        st[j + 4] ^= (~bc[0]) & bc[1];
+
+        j = 10;
+        bc[0] = st[j    ];
+        bc[1] = st[j + 1];
+
+        st[j    ] ^= (~st[j + 1]) & st[j + 2];
+        st[j + 1] ^= (~st[j + 2]) & st[j + 3];
+        st[j + 2] ^= (~st[j + 3]) & st[j + 4];
+        st[j + 3] ^= (~st[j + 4]) & bc[0];
+        st[j + 4] ^= (~bc[0]) & bc[1];
+
+        j = 15;
+        bc[0] = st[j    ];
+        bc[1] = st[j + 1];
+
+        st[j    ] ^= (~st[j + 1]) & st[j + 2];
+        st[j + 1] ^= (~st[j + 2]) & st[j + 3];
+        st[j + 2] ^= (~st[j + 3]) & st[j + 4];
+        st[j + 3] ^= (~st[j + 4]) & bc[0];
+        st[j + 4] ^= (~bc[0]) & bc[1];
+
+        j = 20;
+        bc[0] = st[j    ];
+        bc[1] = st[j + 1];
+        bc[2] = st[j + 2];
+        bc[3] = st[j + 3];
+        bc[4] = st[j + 4];
```

Comment:
Seems like setting `bc[2]`, `bc[3]`, and `bc[4]` here has no impact on the logic, except that it retains the behavior to ensure it loops with those values set (even though they're just overwritten right at the top of the loop). Was this a safety thing you were thinking, just trying to be extra cautious?

Otherwise I figure a macro just like THETA would fit nicely in this section:

```
#define CHI(j) { \
          bc[0] = st[j    ]; \
          bc[1] = st[j + 1]; \
          st[j    ] ^= (~st[j + 1]) & st[j + 2]; \
          st[j + 1] ^= (~st[j + 2]) & st[j + 3]; \
          st[j + 2] ^= (~st[j + 3]) & st[j + 4]; \
          st[j + 3] ^= (~st[j + 4]) & bc[0]; \
          st[j + 4] ^= (~bc[0]) & bc[1]; \
        }

        CHI( 0);
        CHI( 5);
        CHI(10);
        CHI(15);
        CHI(20);
```